client-spark/common/src/main/java/org/apache/spark/shuffle/writer/DataPusher.java

```diff
@@ -0,0 +1,130 @@
+
+public class DataPusher implements Closeable {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DataPusher.class);
+
+  private final ExecutorService executorService;
+
+  private final ShuffleWriteClient shuffleWriteClient;
+  // Must be thread safe
+  private final Map<String, Set<Long>> taskToSuccessBlockIds;
+  // Must be thread safe
+  private final Map<String, Set<Long>> taskToFailedBlockIds;
+  private String appId;
+  // Must be thread safe
+  private final Set<String> failedTaskIds;
+
+  public DataPusher(ShuffleWriteClient shuffleWriteClient,
+      Map<String, Set<Long>> taskToSuccessBlockIds,
+      Map<String, Set<Long>> taskToFailedBlockIds,
+      Set<String> failedTaskIds,
+      int threadPoolSize,
+      int threadKeepAliveTime) {
+    this.shuffleWriteClient = shuffleWriteClient;
+    this.taskToSuccessBlockIds = taskToSuccessBlockIds;
+    this.taskToFailedBlockIds = taskToFailedBlockIds;
+    this.failedTaskIds = failedTaskIds;
+    this.executorService = new ThreadPoolExecutor(
+        threadPoolSize,
+        threadPoolSize * 2,
+        threadKeepAliveTime,
+        TimeUnit.SECONDS,
+        Queues.newLinkedBlockingQueue(Integer.MAX_VALUE),
+        ThreadUtils.getThreadFactory(this.getClass().getName())
+    );
+  }
+
+  public CompletableFuture<Long> send(AddBlockEvent event) {
+    assert appId != null;
+    return CompletableFuture.supplyAsync(() -> {
+      String taskId = event.getTaskId();
+      List<ShuffleBlockInfo> shuffleBlockInfoList = event.getShuffleDataInfoList();
+      try {
+        SendShuffleDataResult result = shuffleWriteClient.sendShuffleData(
+            appId,
+            shuffleBlockInfoList,
+            () -> !isValidTask(taskId)
+        );
+        putBlockId(taskToSuccessBlockIds, taskId, result.getSuccessBlockIds());
+        putBlockId(taskToFailedBlockIds, taskId, result.getFailedBlockIds());
+      } finally {
+        List<Runnable> callbackChain = Optional.of(event.getProcessedCallbackChain()).orElse(Collections.EMPTY_LIST);
+        for (Runnable runnable : callbackChain) {
+          runnable.run();
+        }
+      }
+      return shuffleBlockInfoList.stream()
+          .map(x -> x.getFreeMemory())
+          .reduce((a, b) -> a + b)
+          .get();
+    }, executorService);
+  }
+
+  private synchronized void putBlockId(
+      Map<String, Set<Long>> taskToBlockIds,
+      String taskAttemptId,
+      Set<Long> blockIds) {
+    if (blockIds == null || blockIds.isEmpty()) {
+      return;
+    }
+    taskToBlockIds.computeIfAbsent(taskAttemptId, x -> Sets.newConcurrentHashSet()).addAll(blockIds);
+  }
+
+  public boolean isValidTask(String taskId) {
+    return !failedTaskIds.contains(taskId);
+  }
+
+  public void setAppId(String appId) {
+    this.appId = appId;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (executorService != null) {
+      try {
+        ThreadUtils.shutdownThreadPool(executorService, 5);
```

Comment:
Why is this value `5`?

Reply:
Do you think I need to make it configure?

Reply:
I think it's unnecessary.